patches/sitecustomize.py

```diff
@@ -1,22 +1,56 @@
 # TODO(rosbo): Remove this once we fix the issue with fastai importing older libcudnn if imported prior to tensorflow
 import tensorflow
 import os
+from google.auth import credentials
+from google.cloud import bigquery
+from google.cloud.bigquery._http import Connection
 
-# Monkey patches BigQuery client creation to use proxy.
-kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
-if kaggle_proxy_token:
-    from google.auth import credentials
-    from google.cloud import bigquery
-    from google.cloud.bigquery._http import Connection
+class KaggleKernelCredentials(credentials.Credentials):
+    def refresh(self, request):
+        print("Calling Kaggle.UserSecrets to refresh token.")
+        # Set self.token and self.expiry here.
+        raise NotImplementedError("Private BigQuery integration is not yet implemented.")
 
-    Connection.API_BASE_URL = os.getenv("KAGGLE_DATA_PROXY_URL")
-    Connection._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = kaggle_proxy_token
+kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
+CONNECTION_BASE_URL = Connection.API_BASE_URL
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
+    bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+    specified_project = kwargs.get('project').lower() if 'project' in kwargs else None
+    # Use Data Proxy if user has specified to use the Kaggle project, or if
+    # there are no connected GCP accounts (to maintain backwards compatibility).
+    if bq_user_jwt is None and specified_project and specified_project != 'kaggle':
+        raise Exception("In order to query a private BigQuery project, please connect a GCP account. "
+                        "Otherwise specify 'kaggle' as the project to use Kaggle's public dataset BigQuery integration.")
+    use_data_proxy = specified_project == 'kaggle' or bq_user_jwt is None
+    if use_data_proxy:
+        if data_proxy_project is None or kaggle_proxy_token is None:
+            # We don't have the data proxy info so leave the bq client unmodified.
+            return bq_client(*args, **kwargs)
+        print("Using Kaggle's public dataset BigQuery integration.")
+        Connection.API_BASE_URL = os.getenv("KAGGLE_DATA_PROXY_URL")
+        Connection._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = kaggle_proxy_token
+        anon_credentials = credentials.AnonymousCredentials()
+        anon_credentials.refresh = lambda *args: None
+        kwargs['project'] = data_proxy_project
+        return bq_client(
+            *args,
+            credentials=anon_credentials,
+            **kwargs)
+    else:
+        Connection.API_BASE_URL = CONNECTION_BASE_URL
+        Connection._EXTRA_HEADERS.pop('X-KAGGLE-PROXY-DATA', None)
```

Comment:
How would this header ever be present?

Reply:
In the same case I described above, if the user instantiated a bigquery client first using the Kaggle proxy (ie. client = bigquery.Client(project='KAGGLE')) and then recreated the client with their own private project. It should not be a common case, but wanted to prevent breakage if they did this. Perhaps I should add a unit test for this case.

Reply:
Resolving this since we're covering it above.